### PR TITLE
make_prg update PR series: 4. make_prg update

### PR DESCRIPTION
**Pull request by leoisl**
Hello,
Proceeding with our PR series, this is the most crucial PR, which is the one that adds the `make_prg update` code itself. In the following, I will describe a general view of the code to help reviewing:

### Class `DenovoVariantsDB`

Class used to parse a [denovo variants file](https://github.com/leoisl/make_prg/blob/update/sample_example/denovo_paths/denovo_paths.txt) and provide an index to easily find the updates to be done to a given locus (through member variable `self.locus_name_to_update_data`). It is mostly composed of methods to parse the file and build objects based on the parsed data.

### Class `DenovoLocusInfo` 

Represents all the information of a denovo locus in a [denovo variants file](https://github.com/leoisl/make_prg/blob/update/sample_example/denovo_paths/denovo_paths.txt), e.g. sample, locus name, the ML path through the locus in that sample, and all denovo variants found for that locus for that sample. It is mainly used internally by `DenovoVariantsDB` to get update data for a given locus.

### Class `UpdateData`

Trivial class used to hold updates to be done in a given node of a locus.

### Class `DenovoVariant`

Represents a denovo variant with a ref and alt, and the chromosome would be a ML path through a locus. The ref can go through a single or multiple nodes of the ML path. If it goes through a single node, we can apply the variant to it and get a new sequence (`get_mutated_sequence()`). If it goes through multiple nodes, we can split this variant into a list of sub-variants WRT how it is distributed along the ML path (`split_variant()`).

### Class `MLPath`

Represents a ML path that is indexed in two spaces (path and PRG spaces) for quick queries by other objects.

### Class `MLPathNode`

Represents nodes of the ML path.

## Source file: `make_prg/recursion_tree.py`

This source file contains the classes required to represent the recursion tree of the PRG explicitly. We have two types of nodes in our recursion tree: `MultiClusterNode` and `SingleClusterNode`. `MultiClusterNode` represents a node in the recursion tree such that its respective MSA has to be clustered in two or more clusters, e.g. a non-match interval with at least two groups of different enough sequences. `MultiClusterNode`s are never leaves. `SingleClusterNode` represents nodes such that the sequences of the respective MSA constitute a single cluster (e.g. when the sequences are too close to each other). It is also used when we still don't know how we will cluster the MSA (as concrete examples, the root is always a `SingleClusterNode`, as we don't know how it will be clustered yet, and children of `MultiClusterNode`s are always `SingleClusterNode`s). Depending on the MSA that the `SingleClusterNode` represents, it can have multiple `MultiClusterNode`s and `SingleClusterNode`s children.

In general, a node in the recursion tree stores a bunch of information that would match the recursive call at that point, e.g. nesting level, the MSA itself, parent node, its children, etc. Many of this information and methods are common between `MultiClusterNode`s and `SingleClusterNode`s, so the commonalities were refactored into a single abstract class `RecursiveTreeNode`. Finally, in order to build the PRG from a node, it is enough to traverse it in a preorder, which is the same order of the calls in a recursive algorithm (see `preorder_traversal_to_build_prg()`). `MultiClusterNode` has some unique behaviours, e.g. it is the only type of node that can cluster, while `SingleClusterNode` are the only type of nodes that can be updated. 

This is an example of the recursion tree of a PRG with two `MultiClusterNode`s (in red) and several `SingleClusterNode`s (in blue). Internal MSAs are not represented. Site 15 is not broken into a `MultiClusterNode` because it is actually composed of a single cluster ([this is the case when sequences are all too different](https://github.com/leoisl/make_prg/blob/36d8d5c2179420ce40bfd32c3ef4538955c02650/make_prg/from_msa/cluster_sequences.py#L255)). Sites 17 and 19 are not broken into `MultiClusterNode`s [because it has too few unique sequences (this is a recent optimisation done by @bricoletc that was ported to this code)](https://github.com/leoisl/make_prg/blob/36d8d5c2179420ce40bfd32c3ef4538955c02650/make_prg/recursion_tree.py#L154-L156).

image

### Class `PrgBuilder`

PRG builder based from a multiple sequence alignment. Similar to the existing `PrgBuilder` class, with a few differences, most notably:
1. It now also owns a MSA aligner (`self.aligner`), so that it can rebuild some alignments during the update command;
2. It has an index on the PRG space (`self.prg_index`);
3. It points to a `SingleClusterNode`, which is the root of the recursion tree of this locus' PRG;
4. It can be deserialised and serialised to disk;

### Class `PrgBuilderZipDatabase`

Represents a collection of `PrgBuilder`s, to be saved to and loaded from a zip file.

### CODE REVIEW

**Comment by mbhall88** [comment]
Sorry, I don't think I can provide a meaningful review of this PR. There's so much new stuff here. I trust your thorough testing on the paper data etc.

**Comment by bricoletc** [comment]
> MultiClusterNode represents a node in the recursion tree such that its respective MSA has to be clustered in two or more clusters, e.g. a non-match interval with at least two groups of different enough sequences. MultiClusterNodes are never leaves. SingleClusterNode represents nodes such that the sequences of the respective MSA constitute a single cluster (e.g. when the sequences are too close to each other). It is also used when we still don't know how we will cluster the MSA (as concrete examples, the root is always a SingleClusterNode, as we don't know how it will be clustered yet, and children of MultiClusterNodes are always SingleClusterNodes). Depending on the MSA that the SingleClusterNode represents, it can have multiple MultiClusterNodes and SingleClusterNodes children.

It's not clear to me what the difference really is. A `SingleClusterNode` may get clustered. Why isn't it? 
> This is an example of the recursion tree of a PRG with two MultiClusterNodes (in red) and several SingleClusterNodes (in blue).
What are the numbers inside the nodes? What's the difference between node 3 and node 2, other than their class membership?

**Comment by bricoletc** [comment]
The initial design of `from_msa` was a contrast to something like `from_vcf`. Now source files like prg_builder.py are under `make_prg`, not `from_msa`. I think the file hierarchy needs revisiting. Either we say `make_prg` will only ever support building from MSA, and call the subcommands `build` (instead of `from_msa`) and `update`, or make build and update dirs under `from_msa`?

**Comment by bricoletc** [comment]
As Michael, I don't feel qualified to review the completely new stuff- DeNovo, MLPaths, and trust your tests, but have left comments on the parts I think I understand/want to clarify. I think the codebase has become much larger and more intricate, but that doesn't mean it could be any simpler than what you've done :')

**Comment by mbhall88** [comment]
> The initial design of `from_msa` was a contrast to something like `from_vcf`. Now source files like prg_builder.py are under `make_prg`, not `from_msa`. I think the file hierarchy needs revisiting. Either we say `make_prg` will only ever support building from MSA, and call the subcommands `build` (instead of `from_msa`) and `update`, or make build and update dirs under `from_msa`?

I agree with this too. It seems like changing `from_msa` to simply `build` is much more future-proof. If we add the ability to build from a VCF, we don't need to add another subcommand. We can just infer the file type internally.

**Comment by leoisl** [comment]
> It's not clear to me what the difference really is. A `SingleClusterNode` may get clustered. Why isn't it?

Yeah, some definitions between these two classes overlap. Will try to itemise the main differences between `SingleClusterNode` and `MultiClusterNode` here:
1. `SingleClusterNode` are usually not clustered. When `SingleClusterNode`s create children, [we ensure they are never going to be reclustered](https://github.com/leoisl/make_prg/blob/36d8d5c2179420ce40bfd32c3ef4538955c02650/make_prg/recursion_tree.py#L223-L227). However, `SingleClusterNode` may get clustered if it is the root or if it is child of a `MultiClusterNode`;
2. Leaves are always `SingleClusterNode`s, never `MultiClusterNode`s;
3. Children of `MultiClusterNode`s are always `SingleClusterNode`s, while `SingleClusterNode`s' children can be both;
4. If some intervals of a `SingleClusterNode` must be clustered, it creates `MultiClusterNode`s children with the associated intervals, which will then be clustered.

We could indeed apply a check [when creating `MultiClusterNode`s children](https://github.com/leoisl/make_prg/blob/36d8d5c2179420ce40bfd32c3ef4538955c02650/make_prg/recursion_tree.py#L80-L94) so that we create a `SingleClusterNode` if the associated subalignment will not get clustered, or a `MultiClusterNode` if it will. Then we can also create a `RootClusterNode` class, which is the only one that might be clustered or not. This will indeed separate well the three classes:
1. `SingleClusterNode` never get clustered, they are a single cluster;
2. `MultiClusterNode` always get clustered;
3. `RootClusterNode` may be both;

I am willing to implement these changes, as these definitions currently exist in the code, but their implementation does not totally satisfy them. They overlap a little bit, and this causes confusion. I will however wait for your answer to check if you think this is a reasonable solution or if you propose an alternative one.

> What are the numbers inside the nodes?

When a node is a leaf, we print their associated PRG string. When it is not, we just print its ID, so we can locate it in a debugging session, for example. In an ideal world, instead of the ID, I would like to print their associated MSA, but that is not doable due to space constraints (only way to do this is to actually output a HTML view or sth like that, where when you click on a non-leaf node, it opens a small dialog with the associated MSA);

> What's the difference between node 3 and node 2, other than their class membership?

They represent different MSAs, node `2` is parent of node `3`, and its MSA include node `3`'s MSAs, as well as node `7`'s.

**Comment by leoisl** [comment]
> The initial design of `from_msa` was a contrast to something like `from_vcf`. Now source files like prg_builder.py are under `make_prg`, not `from_msa`. I think the file hierarchy needs revisiting. Either we say `make_prg` will only ever support building from MSA, and call the subcommands `build` (instead of `from_msa`) and `update`, or make build and update dirs under `from_msa`?

> I agree with this too. It seems like changing `from_msa` to simply `build` is much more future-proof. If we add the ability to build from a VCF, we don't need to add another subcommand. We can just infer the file type internally.

I agree with changing `from_msa` to simply `build` and inferring the file type. Would this be fine? Also, could I leave this to an immediate PR after merging this, as it would cause a huge number of changes for just changing a bit the directory/module structure (unsure if git will understand we just reorganised the structure, or will put the changes as large deletions/insertions)?

**Comment by mbhall88** [comment]
> I agree with changing from_msa to simply build and inferring the file type. Would this be fine? Also, could I leave this to an immediate PR after merging this, as it would cause a huge number of changes for just changing a bit the directory/module structure (unsure if git will understand we just reorganised the structure, or will put the changes as large deletions/insertions)?

Fine by me

**Comment by leoisl** [comment]
Sorry for the long delay. https://github.com/iqbal-lab-org/make_prg/pull/36/commits/5bcf9f847ed795c72e848b49021013780221e46d describes a heavy refactoring of the code we use to represent the recursion tree explicitly. This is based on a meeting between @bricoletc and me, that we realised that the code was complicated to read and understand, and thus hard to maintain and error prone. This refactoring touches mostly the classes in `make_prg/recursion_tree.py`. Now we have 5 classes in it:
* `RecursiveTreeNode` : this is an abstract base class for the nodes of the recursion tree, abstracting its common attributes and methods;
* `MultiIntervalNode` : Represent nodes that are clustered vertically, i.e. in intervals;
* `MultiClusterNode` : Represent nodes that are clustered horizontally, i.e. in sequence clusters;
* `LeafNode` : Represent nodes that are never clustered in either direction. These nodes are the only ones that can get updated and indexed;
* `NodeFactory` : Class responsible to build the different `RecursiveTreeNode`s depending on its alignment;

Basically, the difference between `MultiIntervalNode` and `MultiClusterNode` is just how we build the PRG from their children. Building the PRG of a `LeafNode` is basically describing the sequences themselves it represents, as it has no children. Besides, `LeafNode` has update capabilities. All the knowledge on which node to build (i.e. how to cluster the alignment - vertically, horizontally or no cluster) is concentrated on the `NodeFactory` class.

**Comment by leoisl** [comment]
@bricoletc I couldn't re-request a review from you, so I removed and readded you as a reviewer. I hope you got a notification or sth like that, but I am pinging you here just in case...

File: make_prg/recursion_tree.py
Comment (bricoletc) [reply make_prg/recursion_tree.py:135]: I know it's annoying but i think we should use clustering to refer to only horizontal partitioning (kmeans clustering)

File: make_prg/recursion_tree.py
Comment (bricoletc) [inline make_prg/recursion_tree.py:159]: this is repeated with lines 176-177 below, could write a single dedented copy of it at the end

File: make_prg/recursion_tree.py
Comment (bricoletc) [inline make_prg/recursion_tree.py:304]: This does mostly right thing, but not in a clear way IMO

I propose this instead:
```
        min_match_length = prg_builder.min_match_length
        if parent_node is None:
            nesting_level = 0
        else:
            nesting_level = parent_node.nesting_level

        if isinstance(parent_node, MultiIntervalNode):
            clustering_result = kmeans_cluster_seqs(alignment, min_match_length)
            cluster_further = NodeFactory._infer_if_we_should_cluster_further(alignment, clustering_result)
            cluster_further &= (nesting_level + 1 < max_nesting)

            if cluster_further:
                cluster_subalignments = NodeFactory._get_subalignments_by_clustering(alignment, clustering_result)
                result = MultiClusterNode(nesting_level + 1, alignment, parent_node, prg_builder,
                                        cluster_subalignments)
            else:
                result = LeafNode(nesting_level, alignment, parent_node, prg_builder)
        elif isinstance(parent_node, MultiClusterNode):
            all_intervals, match_intervals = NodeFactory._get_all_and_match_intervals(alignment, min_match_length)

            has_a_single_interval = len(all_intervals) == 1
            if not has_a_single_interval:
                interval_subalignments = NodeFactory._partition_alignment_into_interval_subalignments(alignment,
                result = MultiIntervalNode(nesting_level, alignment, parent_node, prg_builder, interval_subalignments)
            else:
                result = LeafNode(nesting_level, alignment, parent_node, prg_builder)
        else:
            raise ValueError("Unsupported parent type for building")
        return result
```

I've made a lot of changes here, here's my rationale:
* Your rules for building MultiInterval or MultiClusterNodes were based on exclusion, but AFAIR if parent is MultiInterval, we attempt horizontal partitioning, and if parent is MultiClusterNode, we attempt vertical partitioning, and that's that. This code reflects this.
* nesting level: needs only to be incremented when building a MultiClusternode. Needs only to be tested for > max_nesting when building a MultiClusterNode
* function `_infer_if_has_single_interval` is weird. it doesnt infer if vertical partition has a single interval, it also tests for max nesting and size of alignment. max nesting should be tested elsewhere (see above point), and alignment size is captured by `IntervalPartitioner` I believe. thus i propose removing this function
* I  propose removing function `_get_nesting_level`, I think code above captures all cases

File: make_prg/recursion_tree.py
Comment (bricoletc) [inline make_prg/recursion_tree.py:341]: I propose removing this, see comment above

File: make_prg/recursion_tree.py
Comment (bricoletc) [inline make_prg/recursion_tree.py:361]: This really does not work IMO. I propose removing it and 'absorbing' it in the code proposed above

File: make_prg/recursion_tree.py
Comment (bricoletc) [inline make_prg/recursion_tree.py:376]: I think this should just be done in `NodeFactory.build()` and not be a function, but leave it to your choice

File: make_prg/recursion_tree.py
Comment (bricoletc) [reply make_prg/recursion_tree.py:304]: The code is a replacement for the whole build() function

**Review by bricoletc — COMMENTED** [review]
I've just made comments about the documentation + proposed a new `build()` which I think is much simpler and perhaps corrects some of the logic, removing the need for support functions in `NodeFactory`. However, I may have missed something/reason why they are there, as I didn't review the update code nor the tests, as update code is beyond me at this time and i trust your tests!

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: It is not, it is only called once... I think you are looking at the code before the refactoring. Github is marking the commented code as outdated, and this is also the updated code: https://github.com/iqbal-lab-org/make_prg/pull/36/files#diff-3a1a023f00c5d0cfacb6033b406775cc18f03701dd9f61b1067dc3eb5666e5bbR294-R296

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: This is a comment also on outdated code, but it is also valid for the updated code. However, I don't actually understand this. This method is [annotated as abstract in the base class](https://github.com/leoisl/make_prg/blob/70c1eb3a4ea82eebebd516bad7effc474efb0261/make_prg/recursion_tree.py#L52-L54) and implemented on concrete subclasses. I don't understand how abstract methods relate to visibility (e.g. private/shared), as they can be both abstract and private, or abstract and shared. I am a bit confused about this, but we can chat about it tomorrow

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: This comment is on outdated code, this method does not exist anymore...

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: This comment is on outdated code

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: This comment is on outdated code, this code does not exist anymore

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: This comment is on outdated code, this `if` does not exist anymore

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: This comment is on outdated code but also applies to the current code. We could indeed do this, but I like to explicitly list the arguments for documenting purposes. If you have a heavy preference for the shortcut version, then we can switch to it

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py:304]: > Your rules for building MultiInterval or MultiClusterNodes were based on exclusion, but AFAIR if parent is MultiInterval, we attempt horizontal partitioning, and if parent is MultiClusterNode, we attempt vertical partitioning, and that's that. This code reflects this.

That works except when the parent is `None`, i.e. the root. The root can be either a `MultiIntervalNode`, a `MultiClusterNode` or a `LeafNode`, and we need to try this order for the root, as we first try to cluster vertically and then horizontally (if it is a `MultiIntervalNode`, we build it. Otherwise, if it is a `MultiClusterNode` we build it. If not, then the root is a leaf). I actually think this exclusion-based rationale is not the most intuitive, but it works for all types of nodes being built. For us to add your changes, we will need a method to build the root, that tries to build it as the 3 types of nodes in that order, which is basically the current implementation

I think the proposed code bugs when we have a single nonmatch interval, e.g.:
```
>1
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
>2
AAAAAAAAAAAAAAAATAAAAAAAAAAAAAAAAA
>3
GGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGGG
>4
GGGGGGGGGGGGGGGGGGCGGGGGGGGGGGGGGG
```
In this case, we can't build a `MultiIntervalNode`, but we can build a `MultiClusterNode` and we should do it, instead of building a `LeafNode`

> nesting level: needs only to be incremented when building a MultiClusternode. Needs only to be tested for > max_nesting when building a MultiClusterNode

I like this suggestion. I have an example where the PRGs built with the current and the suggested code will be different, but I think is fine as they encode the same information... to investigate further, we can talk through this example.

> function _infer_if_has_single_interval is weird. it doesnt infer if vertical partition has a single interval, it also tests for max nesting and size of alignment. max nesting should be tested elsewhere (see above point), and alignment size is captured by IntervalPartitioner I believe. thus i propose removing this function

I agree with this suggestion, although we should take care when we have a single mismatch interval

> I propose removing function _get_nesting_level, I think code above captures all cases

For the suggested code, I agree

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py:376]: I will leave as is because is much easier to test, i.e. it is much easier to mock a function call than mocking several calls to the slice method of biopython's `MultipleSeqAlignment` class... Also, is better documentation than trying to read what the list comprehension does

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py:159]: The one in lines 176-177 is inside a loop, is done several times, not only once

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py:31]: The idea of having a `NodeFactory` class is to delegate the logic of building nodes to it. If we receive the parent node as parameter, then the logic of building the node will have to be put inside the `RecursiveTreeNode` class. This would be clearer in `c++` I think, where I would put this constructor as a private method, and `NodeFactory` as friend of the class, i.e. `NodeFactory` is the only class that is able to construct nodes. This `__init__()` method should never be called by external users, only `NodeFactory.build()` should be called, and all the logic to build nodes is inside `NodeFactory.build()`, this `__init__()` function is just a trivial constructor

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py:364]: Agreed

File: make_prg/recursion_tree.py
Comment (bricoletc) [reply make_prg/recursion_tree.py:159]: yep sorry!

File: make_prg/recursion_tree.py
Comment (bricoletc) [reply make_prg/recursion_tree.py:376]: i see and agree

File: make_prg/recursion_tree.py
Comment (bricoletc) [reply make_prg/recursion_tree.py:31]: In  that case, 
```Builds a new tree node, and its children, recursively, via `NodeFactory````?

Comment (leoisl) [reply make_prg/recursion_tree.py:31]: ouch, I misunderstood your comments, sorry! I agree with them, will make the changes!

File: make_prg/recursion_tree.py
Comment (bricoletc) [reply make_prg/recursion_tree.py:304]: * About your exclusion-based code: i see, I overlooked the root case, and now understand the logic. 
 Could we not then, do
```
if parent_node is None or isinstance(parent_node, MultiClusterNode):
   <same as above>
elif isinstance(parent_node, MultiIntervalNode):
  <same as above>
```
AFAICT at the first `build()`, for the root, we should simply try to vertical partition, and that code above will produce a `MultiIntervalNode` or a `Leaf`

* About the single nonmatch interval, yep I totally forgot about that. (would your tests have picked that up btw?)
So then can we do this?
```
has_a_single_match_interval = len(all_intervals) ==  len(match_intervals) == 1
```

File: make_prg/update/MLPath.py
Comment (leoisl) [reply make_prg/update/MLPath.py:7]: Done in https://github.com/iqbal-lab-org/make_prg/pull/36/commits/75f563f13308e73b8ee9b75f3863381e2e29c1e6 and https://github.com/iqbal-lab-org/make_prg/pull/36/commits/b9c15a1e81ef6704581c679b7b7f23a72b764b49

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: Agreed, done in https://github.com/iqbal-lab-org/make_prg/pull/36/commits/b9c15a1e81ef6704581c679b7b7f23a72b764b49

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: Agreed, done in https://github.com/iqbal-lab-org/make_prg/pull/36/commits/b9c15a1e81ef6704581c679b7b7f23a72b764b49

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: Now that I understood your comments, I implemented your original suggestion: `Builds a new tree node, and its children, recursively` in https://github.com/iqbal-lab-org/make_prg/pull/36/commits/b9c15a1e81ef6704581c679b7b7f23a72b764b49

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: Agreed, done in https://github.com/iqbal-lab-org/make_prg/pull/36/commits/b9c15a1e81ef6704581c679b7b7f23a72b764b49

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: Done in https://github.com/iqbal-lab-org/make_prg/pull/36/commits/3772524ad6d8f7864b9043f973bd9c1d65ec0e80

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: done

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: done

File: make_prg/recursion_tree.py
Comment (leoisl) [reply make_prg/recursion_tree.py]: done

**Comment by leoisl** [comment]
Sorry for the huge delay @bricoletc - I got busy with roundhound as priority. I have finished this some weeks ago, but I am just pushing it now: https://github.com/iqbal-lab-org/make_prg/pull/36/commits/4298300cd713562fd6003102ad2671a6bfa0028b . If you are busy or if I took too long to fix, feel free to approve and we move on

**Review by bricoletc — APPROVED** [review]

**Pull request merged**
